### PR TITLE
Added tools/slo-rules-generator

### PR DESCRIPTION
**Pull request by david-vavra**
Adding tool to generate SLO recording rules with metadata for domains.

The original motivation was adding a new recording rule - `slo:burn_rate_threshold` in addition to already existing `slo:stable_version` and `slo:violation_ratio_threshold`. Introduction of SLO domain configuration lets us to automatically generate the metrics mentioned above and allows us further modify burn_rate alerts on per-domain basis.

*In future we will probably be generating custom burn-rate alerts for each domain,class,type instead of customizing the alerts via recording rules, but we are not there yet and this PR basically just sync our current internal state of tooling to github :) *

### CODE REVIEW

File: tools/slo-rules-generator/slo-domains.yaml.example [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    # 3d: 1
+    # class and slo_type in conditions may be an empty string - which equals to effectively any domain or slo_type
+    burn_rate_thresholds:
+      - condition: {class: 'critical', slo_type: 'availability', time_range: '1h'}
+        value: 13.44
+      - condition: {class: 'low',      slo_type: 'latency90',    time_range: '6h'}
+        value: 33.12
+  classes:
+    critical: # slo_class name
+      # availability slo_type
```
Comment (FUSAKLA) [inline tools/slo-rules-generator/slo-domains.yaml.example:21]:
```suggestion
```

It doesn't say anything new and just complicates the example

File: tools/slo-rules-generator/slo-domains.yaml.example [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+      availability: {slo_threshold: 0.99}
+      # latency90 slo_type
```
Comment (FUSAKLA) [inline tools/slo-rules-generator/slo-domains.yaml.example:23]:
```suggestion
```

File: tools/slo-rules-generator/slo-domains.yaml.example [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+      latency90:    {
```
Comment (FUSAKLA) [inline tools/slo-rules-generator/slo-domains.yaml.example:24]: I really don't get the pattern when the YAML notation is used and when JSON :sweat_smile:

File: tools/slo-rules-generator/slo-domains.yaml.example [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        slo_threshold: 0.90,
+        # k: v which are added as a labels to resulting slo:violation_ratio_threshold metrics
```
Comment (FUSAKLA) [inline tools/slo-rules-generator/slo-domains.yaml.example:26]:
```suggestion
        # labels added to resulting slo:violation_ratio_threshold metrics
```

File: tools/slo-rules-generator/slo-domains.yaml.example [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+        slo_threshold_metadata: {percentile: 90, le: 0.8}
```
Comment (FUSAKLA) [inline tools/slo-rules-generator/slo-domains.yaml.example:27]:
```suggestion
        slo_threshold_metadata: {percentile: 90, le: 0.8} # The 0.8 is a maximum duration of event to be considered successfull taken from the slo_rules.yaml
```

File: tools/slo-rules-generator/slo-rules-generator.go [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+	for k, v := range with {
+		out[k] = v
+	}
+	return out
+}
+
+type SloConfiguration map[string]Domain
+
+func main() {
+	if len(os.Args) != 2 {
```
Comment (FUSAKLA) [inline tools/slo-rules-generator/slo-rules-generator.go:42]: Feature improvement:
Use [Kingpin](https://github.com/alecthomas/kingpin) which simply solves all of this

**Review by FUSAKLA — APPROVED** [review]
Great, generally OK from me

There are a plenty of things we can iterate on, including generating of the alerts etc. leaving out all the joining but for now this is GTG

I added some minor nits for the future improvements, but I'm ok with merging it as is and iterating later.

:+1:

**Comment by FUSAKLA** [comment]
Also, I'd consider mentioning it in the CHANGELOG marked as experimental just to let people know what direction we are heading

File: tools/slo-rules-generator/slo-domains.yaml.example [hunk 7]
```diff
@@ -0,0 +1,31 @@
+userportal-reports: # domain-name
+  enabled: true
+  namespace: production
+  version: 6
+  alerting:
+    team: team.x@company.com
+    escalate: sre.x@company.com
```
[the remark this answers is not in the document]
Comment (david-vavra) [reply tools/slo-rules-generator/slo-domains.yaml.example:7]: Could be...

**Comment by david-vavra** [comment]
Thanks for the review @FUSAKLA, I'm merging it in order to move this forward.

I agree that remaining comments should be addressed, however leaving them to a future refactor.

**Pull request merged**